hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/nio/IOVector.java

```diff
@@ -16,64 +16,70 @@
 
 package com.hazelcast.internal.tpc.nio;
 
-import com.hazelcast.internal.tpc.iobuffer.IOBuffer;
+import com.hazelcast.internal.tpc.buffer.Buffer;
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.channels.SocketChannel;
+import java.util.Arrays;
 import java.util.Queue;
 
+import static com.hazelcast.internal.tpc.util.BitUtil.nextPowerOfTwo;
+
 /**
  * Contains logic to do vectorized I/O (so instead of passing a single buffer, an array of buffer is passed to socket.write).
  */
 public final class IOVector {
 
     private static final int IOV_MAX = 1024;
 
-    private final ByteBuffer[] array = new ByteBuffer[IOV_MAX];
-    private final IOBuffer[] bufs = new IOBuffer[IOV_MAX];
-    private int size;
+    private final Buffer[] buffers = new Buffer[IOV_MAX];
+    private ByteBuffer[] byteBuffers = new ByteBuffer[IOV_MAX];
+
+    private int ioBuffersSize = 0;
+    private int byteBuffersSize = 0;
     private long pending;
 
     public boolean isEmpty() {
-        return size == 0;
+        return ioBuffersSize == 0;
     }
 
-    public void populate(Queue<IOBuffer> queue) {
-        int count = IOV_MAX - size;
+    public void populate(Queue<Buffer> queue) {
+        int count = IOV_MAX - ioBuffersSize;
         for (int k = 0; k < count; k++) {
-            IOBuffer buf = queue.poll();
+            Buffer buf = queue.poll();
             if (buf == null) {
                 break;
             }
-
-            ByteBuffer buffer = buf.byteBuffer();
-            array[size] = buffer;
-            bufs[size] = buf;
-            size++;
-            pending += buffer.remaining();
+            add0(buf);
         }
     }
 
-    public boolean offer(IOBuffer buf) {
-        if (size == IOV_MAX) {
+    public boolean offer(Buffer buf) {
+        if (ioBuffersSize == IOV_MAX) {
             return false;
         } else {
-            ByteBuffer buffer = buf.byteBuffer();
-            array[size] = buffer;
-            bufs[size] = buf;
-            size++;
-            pending += buffer.remaining();
+            add0(buf);
             return true;
         }
     }
 
+    private void add0(Buffer buf) {
+        buffers[ioBuffersSize] = buf;
+        ioBuffersSize++;
+        ByteBuffer[] chunks = buf.getChunks();
+        ensureRemaining(chunks.length);
+        System.arraycopy(chunks, 0, byteBuffers, byteBuffersSize, chunks.length);
```

Comment:
Can you explain why we are copying here? Can't we just write from chunks into the channel?

Reply:
Sorry, but I didn't understand you. Which channel?

Reply:
So we copy from `chunks` to the `byteBuffers` here. And then use `byteBuffers`  to write into the socket channel in `write(SocketChannel)`. The question is, why do we need `byteBuffers` array?

Reply:
Ah, yeah. To be honest, I don't know(it's not my code), but it's a good catch to enhance it.

Reply:
Maybe its better to release buffers instantly.

Reply:
It seems I misunderstand this code a little. I thought we were copying the whole data into different bytebuffers. But we are only copying chunk pointers. To me this is acceptable tradeoff for having a readable code, so I will resolve it.